storage: factor out side effects in preparation of proposer-evaluated KV

tbg opened the pull request:
Systematically go through all the Raft-serialized command implementations and move all of the "side effects" (i.e. actions which must only be performed if the associated batch commits) into the post-commit trigger.

This doesn't get us all the way to proposer-eval'ed KV, but it should hopefully give an idea of what it would continue like. The goal here was to do a fairly 1:1 translation, which explains why the collection of triggers makes for an odd compilation.
~~Some tests are currently disabled and hacks in place, but I am currently working on fixing those as this change is being reviewed.~~Tests are a go.

Review: I'll squash all of this before merge and will add fixes in new commits. The individual commits I think will still help reviewing (if you don't want them, just review the total diff) as they mostly move individual side effects up. However, the downside is that not everything is entirely coherent. When in doubt, just leave a comment.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/8173)

<!-- Reviewable:end -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 7 files at r5, 1 of 4 files at r12, 1 of 2 files at r13, 1 of 3 files at r16, 1 of 1 files at r17.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[storage/replica.go, line 1919 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8173#-KO5r43ayUzlViddcVJL:-KO5r43bfmHMU-ARKtnG:-1128021291) ([raw file](https://github.com/cockroachdb/cockroach/blob/57a4a998b2abfbd8f620b4493314034e093e5283/storage/replica.go#L1919)):*
> ```Go
> }
> 
> const hackIgnore = "hack-ignore"
> ```

Move to the top of the file and add a comment. Unless this is going to be removed soon, give it a better name (`skipTestingCommandFilterID`?)

---

*[storage/replica.go, line 1970 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8173#-KO68KXRWgLgSGb4ezIL:-KO68KXSseRsawutnr-r:-1056372613) ([raw file](https://github.com/cockroachdb/cockroach/blob/57a4a998b2abfbd8f620b4493314034e093e5283/storage/replica.go#L1970)):*
> ```Go
> 			originReplica, ba)
> 		if propResult.PostCommitTrigger != nil {
> 			// Provisionally double-execute the command, but only applying it
> ```

I don't understand this comment.

---

*[storage/replica.go, line 1994 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8173#-KO68WBPWcmKOEqg5Gb3:-KO68WBPWcmKOEqg5Gb4:-121595751) ([raw file](https://github.com/cockroachdb/cockroach/blob/57a4a998b2abfbd8f620b4493314034e093e5283/storage/replica.go#L1994)):*
> ```Go
> 	// remaining writes are the raft applied index and the updated MVCC stats.
> 	//
> 	// TODO(tschottdorf): Considered too dangerous for now.
> ```

Why does this change now? File an issue with your concerns so we'll know what we'll need to be able to re-enable the distinct batch.

---

*[storage/replica.go, line 2035 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8173#-KO69NNWoUs1HmrIqBie:-KO69NNXU7kcF34-tSZ-:1039777109) ([raw file](https://github.com/cockroachdb/cockroach/blob/57a4a998b2abfbd8f620b4493314034e093e5283/storage/replica.go#L2035)):*
> ```Go
> 		r.mu.state.LeaseAppliedIndex = leaseIndex
> 		r.mu.state.Stats = newMS
> 		if forcedError != nil {
> ```

Do we need this condition? I know it was there before but I don't see why we wouldn't want to assert the state here no matter what.

---

*[storage/replica.go, line 2622 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8173#-KO69hvmfX7kWPBID3-p:-KO69hvmfX7kWPBID3-q:-432857521) ([raw file](https://github.com/cockroachdb/cockroach/blob/57a4a998b2abfbd8f620b4493314034e093e5283/storage/replica.go#L2622)):*
> ```Go
> // r.store.metrics. Errors which happen between committing a batch and sending
> // a stats delta from the store are going to be particularly tricky and the
> // best bet is to not have any of those.
> ```

Corruption errors should be rare so we may want the store to just recompute its stats in the background when one occurs.

---

*[storage/replica_command.go, line 807 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8173#-KO6AmLlwVmI7TlqBb_c:-KO6AmLlwVmI7TlqBb_d:1559756176) ([raw file](https://github.com/cockroachdb/cockroach/blob/57a4a998b2abfbd8f620b4493314034e093e5283/storage/replica_command.go#L807)):*
> ```Go
> 		if ct.GetSplitTrigger() != nil {
> 			var err error
> 			if *ms, trigger, err = r.splitTrigger(ctx, batch, *ms, ct.SplitTrigger, txn.Timestamp); err != nil {
> ```

Why doesn't splitTrigger use the same updateTrigger pattern as the others updated in this commit?

---

*[storage/replica_trigger.go, line 45 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8173#-KO6KDljEjHT2OYXrBQ2:-KO6KDljEjHT2OYXrBQ3:-1065952005) ([raw file](https://github.com/cockroachdb/cockroach/blob/57a4a998b2abfbd8f620b4493314034e093e5283/storage/replica_trigger.go#L45)):*
> ```Go
> 	// Store must learn about this delta to update its counters appropriately.
> 	RightDeltaMS enginepb.MVCCStats
> 	// See TODO in the one use of this field.
> ```

Looks like this is set but never read.

---

*[storage/replica_trigger.go, line 171 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8173#-KO6HgpLXjHFTMVI-EM4:-KO6HgpME1pN2H_xUUiE:535755735) ([raw file](https://github.com/cockroachdb/cockroach/blob/57a4a998b2abfbd8f620b4493314034e093e5283/storage/replica_trigger.go#L171)):*
> ```Go
> }
> 
> func (r *Replica) computeChecksumTrigger(
> ```

I think I'd prefer for `{compute,verify}ChecksumTrigger` to live in `replica_command.go` alongside `{Compute,Verify}Checksum`. I think that's better than spreading the checksumming logic across the two files.

---

*[storage/replica_trigger.go, line 355 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8173#-KO6KqBbDnWtROmsk67X:-KO6KqBcYYCw1IZbeein:341305420) ([raw file](https://github.com/cockroachdb/cockroach/blob/57a4a998b2abfbd8f620b4493314034e093e5283/storage/replica_trigger.go#L355)):*
> ```Go
> 		}
> 	}
> 	if trigger.lease != nil {
> ```

This is too big to inline here; pull it out into a separate method. (and maybe put it in `replica_range_lease.go`?)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8173#-:-KO6LIZ56a5x8vcczSIJ:1417088721)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Done. I'll squash-merge this in +12h. Thanks for the review.

---

Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[storage/replica.go, line 1919 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8173#-KO5r43ayUzlViddcVJL:-KO6mWgIC7d9aPRppzcW:1887331193) ([raw file](https://github.com/cockroachdb/cockroach/blob/57a4a998b2abfbd8f620b4493314034e093e5283/storage/replica.go#L1919)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Move to the top of the file and add a comment. Unless this is going to be removed soon, give it a better name (`skipTestingCommandFilterID`?)
</details>
Removed in latest change set.

---

*[storage/replica.go, line 1970 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8173#-KO68KXRWgLgSGb4ezIL:-KO6nu6HJLI9UZJ54pw3:1218506404) ([raw file](https://github.com/cockroachdb/cockroach/blob/57a4a998b2abfbd8f620b4493314034e093e5283/storage/replica.go#L1970)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> I don't understand this comment.
</details>
That's because it was outdated. This is gone and we're now locking via the new `trigger.noConcurrentReads` which is set on splits or merges.

---

*[storage/replica.go, line 1994 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8173#-KO68WBPWcmKOEqg5Gb3:-KO6p2Q1fhPSulTXr8RF:-1041801056) ([raw file](https://github.com/cockroachdb/cockroach/blob/57a4a998b2abfbd8f620b4493314034e093e5283/storage/replica.go#L1994)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Why does this change now? File an issue with your concerns so we'll know what we'll need to be able to re-enable the distinct batch.
</details>
Re-enabled it. I think what I was thinking when I disabled it wasn't correct.

---

*[storage/replica.go, line 2035 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8173#-KO69NNWoUs1HmrIqBie:-KO6pBgpe4FJgVeU0Yxm:1072633123) ([raw file](https://github.com/cockroachdb/cockroach/blob/57a4a998b2abfbd8f620b4493314034e093e5283/storage/replica.go#L2035)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Do we need this condition? I know it was there before but I don't see why we wouldn't want to assert the state here no matter what.
</details>
It's merely a device to make this assertion less expensive. Do you prefer we always run this? I suspect that would show in benchmarks.
Added a comment about this condition.

---

*[storage/replica.go, line 2622 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8173#-KO69hvmfX7kWPBID3-p:-KO6pfG4C2RTip6tDHDn:-1466933873) ([raw file](https://github.com/cockroachdb/cockroach/blob/57a4a998b2abfbd8f620b4493314034e093e5283/storage/replica.go#L2622)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Corruption errors should be rare so we may want the store to just recompute its stats in the background when one occurs.
</details>
Added that to the comment.

---

*[storage/replica_command.go, line 807 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8173#-KO6AmLlwVmI7TlqBb_c:-KO6qbLj_pz7Nz38v23_:-1681983258) ([raw file](https://github.com/cockroachdb/cockroach/blob/57a4a998b2abfbd8f620b4493314034e093e5283/storage/replica_command.go#L807)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Why doesn't splitTrigger use the same updateTrigger pattern as the others updated in this commit?
</details>
Because it was the first one I did. Changed (though it still looks a bit odd).

---

*[storage/replica_trigger.go, line 45 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8173#-KO6KDljEjHT2OYXrBQ2:-KO6r9B9vEcoc5k0AiUQ:177150) ([raw file](https://github.com/cockroachdb/cockroach/blob/57a4a998b2abfbd8f620b4493314034e093e5283/storage/replica_trigger.go#L45)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Looks like this is set but never read.
</details>
Whoops. The corresponding TODO is in the `trigger.split` branch of `handleTrigger`. I've removed this field since it's always true in current usage (splits always recompute everything when something contains estimates).

---

*[storage/replica_trigger.go, line 171 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8173#-KO6HgpLXjHFTMVI-EM4:-KO6rfNRTsyUYz9gdnXr:-1180538916) ([raw file](https://github.com/cockroachdb/cockroach/blob/57a4a998b2abfbd8f620b4493314034e093e5283/storage/replica_trigger.go#L171)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> I think I'd prefer for `{compute,verify}ChecksumTrigger` to live in `replica_command.go` alongside `{Compute,Verify}Checksum`. I think that's better than spreading the checksumming logic across the two files.
</details>
I don't disagree, but I think we yet have to discuss where this type of code should live and how it should be organized. After all, everything in a trigger is more likely to be sensitive to rolling updates, and I feel rather strongly about de-muddling "Raft" and everything else (though checksum computation is some sort of weird hybrid). In the interest of not opening that box in this PR, I'd like to leave it here - I expect @eisenstatdavid to touch (and move) this code sooner than later anyway.

---

*[storage/replica_trigger.go, line 355 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8173#-KO6KqBbDnWtROmsk67X:-KO6tBbk77O5Jdr20w8e:-875415751) ([raw file](https://github.com/cockroachdb/cockroach/blob/57a4a998b2abfbd8f620b4493314034e093e5283/storage/replica_trigger.go#L355)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> This is too big to inline here; pull it out into a separate method. (and maybe put it in `replica_range_lease.go`?)
</details>
Done. I've limited to mechanical extraction in the same file to limit the amount of beautification in this PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8173#-:-KO6t_NUsNPJmx2OPIbe:-1014268371)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
(Of course with an appropriate commit message and only in the absence of commentary)

---

Review status: 2 of 6 files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8173#-:-KO6tiGF7Bc3Y6jXfKEk:724452108)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
I'll also fix the flakiness in `TestRangeTransferLease` I seem to have introduced.

bdarnell commented:
Reviewed 3 of 4 files at r18, 1 of 1 files at r19.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[storage/replica.go, line 2035 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8173#-KO69NNWoUs1HmrIqBie:-KO7kp2g7UA1ZlGExpMc:1162081916) ([raw file](https://github.com/cockroachdb/cockroach/blob/57a4a998b2abfbd8f620b4493314034e093e5283/storage/replica.go#L2035)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> It's merely a device to make this assertion less expensive. Do you prefer we always run this? I suspect that would show in benchmarks.
> Added a comment about this condition.
</details>
Ah, sorry, I read the condition backwards and thought we were running the checks in the common case. Since we're only doing it on error it makes sense to keep the `if`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8173)*
<!-- Sent from Reviewable.io -->

tbg committed:
storage: refactor for proposer-eval'ed kv

Starting from `applyRaftCommand`, refactor in preparation for #6286 (which at
this point is "almost" closed, missing only a second audit and some refactoring
which makes it *impossible* for those methods to do anything but use the
batch).

Due to the subtlety of this change, it should be a prime suspect for anomalies
in the near future. The movement of code may have changed some test semantics
in subtle ways, in the most dangerous case letting them pass while hiding a
regression. However, most of the changes were relatively mechanical and
appropriate care has been taken.

Prepare `applyRaftCommand` for #6290. After that issue, `applyRaftCommand`
will operate on a supplied `WriteBatch` and an associated protobuf containing
information necessary to execute all relevant side effects.

The `WriteBatch` will be constructed on the lease holder (by calling some new
method which will contain much of what is now the part of `applyRaftCommand`
leading down to `executeWriteBatch`).

The heavy lifting here is removing all spurious side effects from the state
machine methods multiplexed to via `(*Replica).executeCmd`.

In particular, in this change we

- remove a first handful of `(*Batch).Defer`s, introduced a type
  `proposalResult` which along with the `WriteBatch` must contain all the
  information necessary for applying the command. Sprinkled TODOs liberally to
  give a general sense of what's to be done.

- hoist communicating MVCCStats delta to store and leave a TODO for the
  handling of `ReplicaCorruptionError`s.

- isolate a maybeGossipSystemConfig trigger

- remove all of the uses of `batch.Defer()` in `replica_command.go`, for
  instance on Split, Merge (required some stats cleanup), GC, TruncateLog,
  ChangeFrozen, RequestLease, TransferLease; lift their side effects
  appropriately as well

- move call to `setDesc` to trigger

- remove Defer in handleRaftReady

- move checksum computation and verification to triggers

- reset ContainsEstimates in split post-commit trigger (in effect moving that
  hack up from splitTrigger)

- move lease success/error counter update to trigger
